`docker squash`: Consolidate image layers

jlhawn opened the pull request:
This patch adds a new CLI command:

```
$ docker squash --help

Usage: docker squash [OPTIONS] IMAGE [ANCESTOR]

Merge filesystem layers of an image into a new image

  --help=false        Print usage
  --no-trunc=false    Don't truncate output
```

Example use cases:
- Consolidating all layer of an image to a new single-layered image.
  
  Let's inspect the history of the `busybox` image.
  
  ```
  $ docker history busybox
  IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
  8c2e06607696        8 weeks ago         /bin/sh -c #(nop) CMD ["/bin/sh"]               0 B                 
  6ce2e90b0bc7        8 weeks ago         /bin/sh -c #(nop) ADD file:8cf517d90fe79547c4   2.43 MB             
  cf2616975b4a        8 weeks ago         /bin/sh -c #(nop) MAINTAINER Jérôme Petazzo     0 B
  ```
  
  It has 3 layers total, but we can now `squash` it and tag it as something else.
  
  ```
  $ docker squash --tag better_busybox busybox
  ccbe958582b2
  ```
  
  Now we have these 2 images.
  
  ```
  $ docker images
  REPOSITORY          TAG                 IMAGE ID            CREATED              VIRTUAL SIZE
  better_busybox      latest              ccbe958582b2        About a minute ago   2.43 MB
  busybox             latest              8c2e06607696        8 weeks ago          2.43 MB
  ```
  
  But `better_busybox` only has a single layer in its history.
  
  ```
  $ docker history better_busybox
  IMAGE               CREATED              CREATED BY          SIZE                COMMENT
  ccbe958582b2        About a minute ago                       2.43 MB 
  ```
- Merging layers of a newly built image into a single layer relative to the base image.
  
  Let's build an image using this Dockerfile:
  
  ``` Dockerfile
  FROM busybox:latest
  
  MAINTAINER Josh Hawn <josh.hawn@docker.com>
  
  ADD docker/docs /docs
  RUN echo hello > /hello.txt
  ```
  
  ```
  $ docker build -t test_build .
  Sending build context to Docker daemon 93.87 MB
  Step 0 : FROM busybox:latest
   ---> 8c2e06607696
  Step 1 : MAINTAINER Josh Hawn <josh.hawn@docker.com>
   ---> Running in 0c2b29e087d5
   ---> 4ac4b75a88f6
  Removing intermediate container 0c2b29e087d5
  Step 2 : ADD docker/docs /docs
   ---> c9e50d91bc69
  Removing intermediate container ad9e921d943f
  Step 3 : RUN echo hello > /hello.txt
   ---> Running in e15283387c37
   ---> 49644b91f444
  Removing intermediate container e15283387c37
  Successfully built 49644b91f444
  ```
  
  And inspect the history of this new image:
  
  ```
  $ docker history test_build
  IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
  49644b91f444        24 minutes ago      /bin/sh -c echo hello > /hello.txt              6 B                 
  c9e50d91bc69        24 minutes ago      /bin/sh -c #(nop) ADD dir:24ae8a736955084e39d   8.537 MB            
  4ac4b75a88f6        24 minutes ago      /bin/sh -c #(nop) MAINTAINER Josh Hawn <josh.   0 B                 
  8c2e06607696        8 weeks ago         /bin/sh -c #(nop) CMD ["/bin/sh"]               0 B                 
  6ce2e90b0bc7        8 weeks ago         /bin/sh -c #(nop) ADD file:8cf517d90fe79547c4   2.43 MB             
  cf2616975b4a        8 weeks ago         /bin/sh -c #(nop) MAINTAINER Jérôme Petazzo     0 B
  ```
  
  See that it has 6 layers. That's 3 more than were already in the base image. We can squash these 3 into an image with one additional layer relative to the base image, and re-tag this new image.
  
  ```
  $ docker squash --tag test_build test_build busybox
  ```
  
  Now `test_build` only has a single new layer on top of the original 3 layers from `busybox`:
  
  ```
  $ docker history test_build
  IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
  97561533b0fa        2 minutes ago                                                       8.537 MB            
  8c2e06607696        8 weeks ago         /bin/sh -c #(nop) CMD ["/bin/sh"]               0 B                 
  6ce2e90b0bc7        8 weeks ago         /bin/sh -c #(nop) ADD file:8cf517d90fe79547c4   2.43 MB             
  cf2616975b4a        8 weeks ago         /bin/sh -c #(nop) MAINTAINER Jérôme Petazzo     0 B
  ```

Commit:
[daemon/graphdriver] Add DiffAncestor method

DiffAncestor is a new method of graphdriver.Driver which produces an archive
of the changes between one layer and an ancestor layer. More generally, the
naiveDiffDriver produces an archive of changes between any two directories.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

Commit:
[graph] Add ability to squash image layers

graph.Squash consolidates all layers in between the given image and an
ancestor image into a single layer and returns the newly created image. A nil
ancestor argument indicates that the layer should be squashed all the way to
a single new base image.

tagStore.Squash first performs a lookup of a given image name or ID and the
ancestor name or ID before calling graph.Squash. This allows you to use
image tags rather than IDs to squash image layers.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

thaJeztah commented:
Thanks @jlhawn! before we proceed, I'm interested to hear if @crosbymichael agrees with this implementation, based on https://github.com/docker/docker/issues/332#issuecomment-102133413

> For relieving the immediate pain, we also suggest updating the documentation on how to manually squash layers if people really need it, with external tools. @crosbymichael volunteered on writing such a tool.

ping @crosbymichael interested in your opinion, to prevent giving people "false hope"

Commit:
[api/server] Add POST /images/.../squash endpoint

A new API endpoint in needed to remotely request that a new image be created
by consolidating the layers of one image or the layers between an image and
an ancestor image. Image IDs or names/tags are accepted.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

jlhawn commented:
@thaJeztah I'm happy to wait to see what the rest of the team thinks, but I'll go ahead and implement it anyway for the following reasons:

- I think there's way more than just the need to do 'absolute' squashing (which can be done manually today with `docker export` but is kind of a hack). With 'relative' layer squashing you still get all of the benefits of shared rootfs layers.
- I think the functionality can be added to Docker much more elegantly than any external tool because it's already really simple to do with the `graphdriver` interface.
- I'm motivated to do it *now* so I'd rather not wait and potentially be discouraged.
- It's fun.

Commit:
[api/client] Add `docker squash` command

This new CLI command allows users to merge the layers of an image into a
new image, either all images or up to an ancestor layer.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

jsm commented:
@heph

jlhawn commented:
There's something very strange happening with the Jenkins builds and the docker-py tests :-(
I also still need to add some integration-cli tests for this as well as documentation changes of course.

thaJeztah commented:
> I'm motivated to do it now so I'd rather not wait and potentially be discouraged. It's fun.

:+1: don't let me stop you!

Commit:
[integration-cli] Test cases for `docker squash`

Tests absolute and relativ squashing and well as squash conditions which
should result in non-ops returning the same image.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

Commit:
[docs] Document `docker squash` and API changes

Updates API v20 markdown as well as adds manpage markdown for
`docker squash` and adds `squash` to the cli reference markdown.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

jlhawn commented:
I've added integration tests and docs, so this is no longer a work-in-progress.

@vincentwoo @tianon @cpuguy83 @SvenDowideit @thaJeztah @yosifkit @rhatdan please take a look and/or try it out and let me know what you think.

duglin commented on api/client/squash.go:
```diff
@@ -0,0 +1,50 @@
+	cmd.Require(flag.Min, 1)
+	cmd.Require(flag.Max, 2)
+
+	noTrunc := cmd.Bool([]string{"#notrunc", "-no-trunc"}, false, "Don't truncate output")
```
why have a deprecated option  on a new cmd?

jlhawn replied:
oh, is that what the `#notrunc` means? I actually just copied this section from the implementation of the `docker history` command. :disappointed: 

... 

Just got back from reading the comments/docs on `pkg/mflag`. It all makes sense now :dizzy_face:

duglin commented:
Oops - sorry -jumped the gun on the design review.  Wanna squash (LOL) your commits? :-)

vbatts commented:
and the squash to [ancestor] is if you just want to squash back to a particular point in the history?

Also, is the `docker build` related history stored in the squashed layers?

otherwise, i'm glad someone else is _again_ trying to get squash in... and glad it's you.

duglin commented:
If this makes it in, which I hope it does, I think we need to reconsider the Dockerfile SQUASH command. Would be really odd to force people to build and **then** squash instead of doing it all in one step.

tianon commented:
I've got the same question as @vbatts:
> Also, is the `docker build` related history stored in the squashed layers?

If the answer is "yes" (and I'm hoping so), then why not have `docker build` just always squash all the way up to the `FROM` with an optional `--no-squash` flag?

tianon commented:
It'd also be fun to be able to tweak the "squash ancestor" if we do go for squash-on-build, but I suppose that entire discussion is really orthogonal.

jlhawn commented:
@vbatts asked:

> and the squash to [ancestor] is if you just want to squash back to a particular point in the history?

Yes, exactly this. What I wanted to do in #9591 was 'squash' up to the base image but this gives you a little more control over exactly how far you want to go back. Unfortunately there's no way to squash layers that are in the middle of an image with this, i.e., there are 5 layers in my image and i want to combine the middle 3 and end up with 3 total. That would require modifying the parent pointer of the last layer which is something I don't want to do (yet).

and @tianon:

> Also, is the docker build related history stored in the squashed layers?

At the time, no :-( the build related history has always been spread across all layers of an image. If you look at the example `docker history` outputs above for squashed images, the squashed history entries have empty `CREATED BY` and `COMMENT` columns. If there's enough interest, I can combine the `CREATED BY` and `COMMENT` sections of the intermediate layers.

and @duglin:

> Would be really odd to force people to build and then squash instead of doing it all in one step.

These thoughts were also echoed by @tianon, but I agree with him when he says it's orthogonal. Let's get the separate functionality in first then we can add options on the other commands for doing it.

In a similar way, there's no option to tag these new squashed images, they have to be separate. Adding it is relatively simple, like adding the `-t` flag to `docker build` but I think it could be done separately too, just like adding `--squash` or `--no-squash` flags to other commands.

tianon commented:
IMO, this would be much more interesting if it could still tie into the
build cache -- as-is, it's a nice feature, but not one I'd ever see myself
actually using since so much useful metadata is lost.  Concatenating all
the "comment" fields seems like the best solution to me, especially since
that makes "docker history" output stay useful, which is a big +1 from me
(and ties in well to how the build cache works, making adding these
squashed layers into the build cache considerations a somewhat more real
possibility).

vbatts commented:
also, #4232 ..

jlhawn commented:
from https://github.com/docker/docker/pull/4232#issuecomment-59581289:

> @vbatts and the other maintainers are currently working on a new image format that will support these advanced operations without destroying history of how the image was built. We can close this as it will be addressed in the new format.

 this hasn't happened, can't happen, and isn't going to happen in any Docker 1.X timeframe.

thaJeztah commented:
Although I really would like to see this merged. The biggest downside of this PR (compared, e.g. to the `MARK..SQUASH` proposal), is that this *cannot* be used for automated builds, which severely limits it's usefulness.

Being only able to squash the latest x layers, sounds limiting in cases where larger and less-frequently changing layers are at the start of my Dockerfile, and I want to squash those. (committing those as a base-image is probably the only opion there).

I really wish the `MARK..SQUASH` could be reconsidered.

(please don't consider this a negative comment on this PR, I really appreciate this)

jlhawn commented:
> Being only able to squash the latest x layers, sounds limiting in cases where larger and less-frequently changing layers are at the start of my Dockerfile, and I want to squash those. (committing those as a base-image is probably the only opion there).

Right, I mentioned this in my comment above, and having an intermediate Dockerfile and squashing that would be the only way to go, short of having a squash in the middle of your Dockerfile.

I know it's natural to want to mix this feature in with `docker build` and begin thinking about how it can affect automated build processes, but I'd like to consider it separately. If this is accepted, we can certainly add a new Dockerfile directive and have the build cache be aware of squashed layers.

Just because it doesn't immediately benefit those use cases doesn't mean it benefits no one.

thaJeztah commented:
> Just because it doesn't immediately benefit those use cases doesn't mean it benefits no one.

:+1: agreed

duglin commented:
Just from a consistency, and fairness, point of view we need those who were against the Docker build squash feature to speak up here. Some of the reasons that it was rejected were (from what I remember):
- didn't fully understand the need for it so they wanted more focused issues/PR opened
- didn't want to expose layers to users (thought it was more of an internal daemon concept)
- some people thought the performance hit of having large number of layers could/should be solved via something different

As far as I can tell, these would still apply to this feature so its hard not to link this one with the docker build one.  And, in fact, I would much prefer to discuss both at the same time because I would want the solution for both of them to be consistent.  Meaning, if possible, they have similar UX aspects.

jlhawn commented:
@tianon I've updated the `[graph]` commit to keep all of the build commands in the new image layer, as well as comments from each layer.

jlhawn commented:
> - didn't fully understand the need for it so they wanted more focused issues/PR opened
> - didn't want to expose layers to users (thought it was more of an internal daemon concept)
> - some people thought the performance hit of having large number of layers could/should be solved via something different

I understand wanting to abstract away the concept of layers as an implementation detail. This is very much desired in the new image format, but Docker today makes no distinction at all between layers and images. Layers *are* images as far as the Docker daemon is concerned.

Though we try to put up an abstraction barrier by making a distinction between them or not even mentioning the 'layer' concept to new users, it leaks all over the place, in the API and in the CLI, and in the `docker build` process. Layers may be a more advanced and lower-level concept but there are a lot of advanced users who understand it and want more control over it.

`docker squash` doesn't have to be a "Basics of Docker" concept that new users are expected to understand but eventually they may need to, similar to how new users of `git` aren't introduced to the concept of rebasing commits on the first day but it eventually becomes a powerful and commonly used tool.

vincentwoo commented:
@jlhawn thanks for going the distance on this one. I'm glad I pinged you.

As an end user who would start using this immediately, I'd also want to know:

* Can we add an easy `-t` option for tagging? There's no situation in which I'd use this and not immediately tag the result.
* Does this handle whiteouts efficiently? A lot of the charm is I can unroll all my long `RUN` commands with `&&` into individual `RUN`s, but only if the layers that get deleted don't actually end up in the final image.
* Does it handle all the metadata merging, too? `ENV`, `USER`, etc?

jlhawn commented:
> - Can we add an easy -t option for tagging? There's no situation in which I'd use this and not immediately tag the result.

Yeah, I understand that can be pretty common. I should have a little extra time to add this.

> - Does this handle whiteouts efficiently? A lot of the charm is I can unroll all my long RUN commands with && into individual RUNs, but only if the layers that get deleted don't actually end up in the final image.

whiteouts - now there's an implementation detail that shouldn't be exposed!

The efficiency of squashing layers depends solely upon how many files/directories are in the filesystem. It uses the same method of generating a layer diff as all storage drivers do (aside from AUFS) with their parent layer: Mount both the image rootfs and its ancestor and perform a filepath walk of both, detecting all additions, changes, and deletions. The more stuff that there is in either filesystem then the longer time squashing will take. It's really no different than `docker commit` in that respect.

> - Does it handle all the metadata merging, too? ENV, USER, etc?

Yes, the resulting image will have the same runtime config as the original image.

vincentwoo commented:
Ah, what I meant by "efficiently" was just that "removed nodes do not end up in the squashed final image", and less "does the squash go fast". I take it from your answer that they do not, which is dope!

jlhawn commented:
Oh, sorry. Yeah, they will not end up in the resulting image. Keep in mind, though, that squashing *does* require additional disk space (on the machine that builds and squashes the layers) as it effectively copies everything that was in the final image. So you will need to delete the unused original layers to reclaim that disk space.

vincentwoo commented:
Sounds great. There are quite a few things that I think will make this CLI tool better, but they're all pointless unless this gets merged. +1 to getting this in!

rhatdan commented:
I also want to get this in,  We have a home grown version of this running for our internal builds because of the amount of layers required to build some JBOSS applications.  Getting a standard way is a big plus.

jlhawn commented:
@vincentwoo @duglin 

> - Can we add an easy -t option for tagging? There's no situation in which I'd use this and not immediately tag the result.

Done :smiley:

vincentwoo commented:
@jlhawn dockercon's over what's it gonna take to get this bad boy merged

vincentwoo commented:
@jlhawn can you give us an update on the status of this PR? Is it still waiting review?

tiborvass commented:
@jlhawn I think this will have to be reevaluated/rethought once the client-side builder is in place. WDYT? Would you mind closing it?

icecrime commented:
1. We want to move the builder out first
2. Layers are an implementation detail that should not leak in the UX

Sorry, closing this.

vincentwoo commented:
Does that mean you're nixing stuff like `docker history` et al?
